TCD-2901: Backtrace - Apache and Sourcemap (#2152)
saucelabs/sauce-docs: docs/error-reporting/platform-integrations/apache.md

---
id: apache
title: Integrating Apache Traffic Server
sidebar_label: Apache Traffic Server
description: Configure Apache Traffic Server to use Backtrace's Invoker as a crash log helper.
---

import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';
import useBaseUrl from '@docusaurus/useBaseUrl';

> This document provides instructions on how to configure Apache Traffic Server (ATS) to use Backtrace's Invoker as a crash log helper. The workflow upon ATS crash will be as follows:

Author: ```suggestion
This document provides instructions on configuring Apache Traffic Server (ATS) to use Backtrace's Invoker as a crash log helper. The workflow upon ATS crash is as follows:
```

> 1. On startup, ATS will launch the invoker, which will monitor the `traffic_server` process.

Author: ```suggestion
1. ATS launches the invoker on startup, which monitors the `traffic_server` process.
```

> 2. Upon `traffic_server` crash, the invoker will launch `ptrace` via a script that provides metadata.

Author: ```suggestion
2. Upon the `traffic_server` crash, the invoker launches `ptrace` via a script that provides metadata.
```

- (Optional but recommended) - Install the morgue command-line queryingtool: (See: [Morgue Installation](https://help.backtrace.io/product-guide/morgue))

:::tip Generate a Submission Token

1. In the Backtrace Console, go to **Project settings > Error submission > Submission tokens**.
1. Select **+**.

:::

## Set Up Helper Scripts

### `invoker.sh`

```bash
#!/bin/sh
exec /opt/backtrace/bin/invoker -t "/home/ats/backtrace.sh %p" $@
```

### `backtrace.sh`

This script calls `ptrace` with any metadata you wish to gather and then submits it to the coroner. Here, we're extracting the version number using `traffic_ctl`, but feel free to add any additional metadata that you need.

```bash
#!/bin/sh

## Change the following as needed
PROJECT=ats
TOKEN=ats
DM=/export/ats
CF=/etc/coronerd/coroner.cf

PATH=/opt/backtrace/bin:/usr/bin:/bin:$PATH

# Add commands to gather any metadata that you wish to pass to the ptrace invocation
VERSION=`/opt/ts/bin/traffic_ctl metric get proxy.node.version.manager.short|cut -d ' ' -f 2`

mkdir -p ${DM}/
ptrace --kv="version:$VERSION" $1 -O ${DM}/ats
if test "$?" == "0"; then
   coroner -c $CF put -u ${PROJECT} ${TOKEN} ${DM}/*.btt

fi
```

> Ensure that your `traffic_server` processes can execute these scripts and have permission to write to the destination folder specified in `backtrace.sh` (referred to by the variable `DM`).

Author: ```suggestion
Ensure that your `traffic_server` processes can run these scripts and have permission to write to the destination folder specified in `backtrace.sh` (referred to by the variable `DM`).
```
